### PR TITLE
JENKINS-35206 - fix potential NPE due to corrupt Jenkins.InstallState

### PR DESCRIPTION
**Pull request by kzantow**
Somehow `Jenkins.installState` was being corrupted; add some additional checking for this case and just return `InstallState.UNKNOWN`.

@reviewbybees esp. @oleg-nenashev

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
IMHO the only realistic case of getting such state is a corrupted config file with the legacy non-transient <code>installState</code> due to the whatever reason (e.g. upgrade from the previous version). I would vote for handling it in InstallState#readResolve() or in Jenkins#readResolve()

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 1]
```diff
@@ -953,6 +953,9 @@ private Object readResolve() {
     @Nonnull
     @Restricted(NoExternalUse.class)
     public InstallState getInstallState() {
+        if (installState == null || installState.name() == null) {
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/model/Jenkins.java]: This approach is hackish. It won't be extensible if you add new fields to the InstallState

File: core/src/main/java/jenkins/install/InstallUtil.java [hunk 2]
```diff
@@ -93,10 +93,6 @@ public T get() {
      * Proceed to the state following the provided one
      */
     public static void proceedToNextStateFrom(InstallState prior) {
-        InstallState current = Jenkins.getInstance().getInstallState();
-        if (!current.equals(prior)) {
-            if (Main.isDevelopmentMode) LOGGER.warning("Transitioning state from: " + prior + ", but current is: " + current);
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/install/InstallUtil.java]: IMHO it's useful not only in the Development mode. Deletion as a bugfix seems to be a strange approach

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 3]
```diff
@@ -953,6 +953,9 @@ private Object readResolve() {
...
+            return InstallState.UNKNOWN;
```
Comment (recena) [inline core/src/main/java/jenkins/model/Jenkins.java]: Maybe a log message?

File: core/src/main/java/jenkins/install/InstallUtil.java (hunk 2 above)
Comment (kzantow) [inline core/src/main/java/jenkins/install/InstallUtil.java]: Actually, it's possible you might want to transition different stages around, this is not necessarily the right thing to check for.

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 1 above)
Comment (kzantow) [inline core/src/main/java/jenkins/model/Jenkins.java]: Not sure why this is hackish? constructor is defined as: `InstallState(@Nonnull String name...`, only way for installState.name to be null is that it's corrupt, in which case we simply don't know what the state is (hence, `UNKNOWN`), and `proceedToNextStateFrom(UNKNOWN)` will be called.

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 3 above)
Comment (kzantow) [inline core/src/main/java/jenkins/model/Jenkins.java]: I think it's irrelevant, it's been corrupted (by XStream or some other spot), corrupt state is an unknown state.

Comment (recena) [inline core/src/main/java/jenkins/model/Jenkins.java]: @kzantow But it is a special case, right? This cases are candidates to be mentioned in the log system. At least, this is my opinion.

Comment (kzantow) [inline core/src/main/java/jenkins/model/Jenkins.java]: Maybe, but there is absolutely nothing that anyone can do about it here. When did it get corrupted? This gives no information why it's in a bad state.

Comment (recena) [inline core/src/main/java/jenkins/model/Jenkins.java]: @kzantow You decide ;)

**Comment by oleg-nenashev** [comment]
I'm fine with such hardening if we also get https://github.com/jenkinsci/jenkins/pull/2387 integrated.

**Comment by oleg-nenashev** [comment]
🐝 from me since it's harmless
@recena WDYT?

**Comment by recena** [comment]
@oleg-nenashev I shared my opinion, but it is not an error at all. Go ahead!

🐝

**Comment by oleg-nenashev** [comment]
@reviewbybees done

**Pull request merged**